src/v/cluster/archival/archiver_operations_impl.cc

```diff
@@ -901,4 +908,317 @@ ss::shared_ptr<archiver_operations_api> make_archiver_operations_api(
 
 } // namespace detail
 
+struct one_time_stream_provider : public stream_provider {
+    explicit one_time_stream_provider(ss::input_stream<char> s)
+      : _st(std::move(s)) {}
+    ss::input_stream<char> take_stream() override { return std::move(*_st); }
+    ss::future<> close() override {
+        if (_st.has_value()) {
+            return _st->close();
+        }
+        return ss::now();
+    }
+    std::optional<ss::input_stream<char>> _st;
```

Comment:
It's set to `std::nullopt` in the `take_stream` method so the method can only be called once. It's `struct` because it's not exposed outside of the .cc so it's not a big deal. The assumption here is that the object will be constructed and passed to the `remote` which uses it thought the `stream_provider` interface so it won't be able to access `_st` directly.

Reply:
Fair enough.

FYI, `std::move(*_st)` does not make `_st == std::nullopt`, nor does `std::move(_st)`. [Ref](https://en.cppreference.com/w/cpp/utility/optional/optional)